cmd/zoekt-sourcegraph-indexserver/backoff_test.go

```diff
@@ -0,0 +1,249 @@
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	time.Sleep(backoffDuration * 2)
+
+	queue.Bump([]uint32{opts.RepoID})
+
+	if _, ok := queue.Pop(); !ok {
+		t.Fatal("queue should no longer be empty after waiting for longer than the backoff duration and then bumping index options")
+	}
+}
+
+func TestQueue_ResetBackoffUntil(t *testing.T) {
+	backoffDuration := 1 * time.Hour
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	queue.SetIndexed(opts, indexStateSuccess)
+
+	queue.Bump([]uint32{opts.RepoID})
+
+	if _, ok := queue.Pop(); !ok {
+		t.Fatal("queue should no longer be empty after resetting backoff until time to zero value")
+	}
+}
+
+func TestQueue_ResetFailuresCount(t *testing.T) {
+	backoffDuration := 1 * time.Millisecond
+	maxBackoffDuration := 1000 * time.Millisecond
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	// consecutive failures will push backoff until to a further out time
+	for i := 0; i < 1000; i++ {
+		queue.SetIndexed(opts, indexStateFail)
+	}
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	queue.SetIndexed(opts, indexStateSuccess)
+
+	queue.Bump([]uint32{opts.RepoID})
+
+	// backoff until is only one duration in the future after resetting consecutive failures count
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	time.Sleep(backoffDuration)
+
+	queue.Bump([]uint32{opts.RepoID})
+
+	if _, ok := queue.Pop(); !ok {
+		t.Fatal("queue should no longer be empty after waiting a backoff duration for the first failure")
+	}
+}
+
+func TestQueue_IncreaseDurationWithFailuresCount(t *testing.T) {
+	backoffDuration := 5 * time.Millisecond
+	maxBackoffDuration := 1000 * time.Millisecond
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	sleep := 1 * time.Millisecond
+	for i := 0; i < 10; i++ {
+		queue.SetIndexed(opts, indexStateFail)
+
+		if _, ok := queue.Pop(); ok {
+			t.Fatal("queue should be empty after SetIndexed")
+		}
+
+		// bump should have no impact on queue during backoff duration which increases with consecutive failures count
+		time.Sleep(sleep)
```

Comment:
this can be flakey

Reply:
We could discard this test and cover it in a backoff struct unit test

Reply:
This test is removed